amxmodx/file.cpp

```diff
@@ -7,484 +7,503 @@
 			}
-			//int a, b = '\n';
-			//while( (a = fgetc(fp)) != EOF ){
-			//  if ( a == '\n')
-			//    ++lines;
-			//  b = a;
-			//}
-			//if ( b != '\n' )
-			//  ++lines;
+
 			return lines;
 		}
-		else if (params[2] == 2)
+		case FSOPT_END_WITH_LF:
 		{
-			fseek(fp, -1, SEEK_END);
-			
-			if (fgetc(fp) == '\n')
-				return 1;
-			
-			return 0;
+			int8_t ch = 0;
+
+			fp->Seek(-1, SEEK_END);
+			fp->Read(&ch, 1);
+
+			return ch == '\n';
 		}
 	}
-	
+
 	return -1;
 }
 
+// native fopen(const filename[], const mode[], bool:use_valve_fs = false, const valve_path_id[] = "GAME");
 static cell AMX_NATIVE_CALL amx_fopen(AMX *amx, cell *params)
 {
-	int len;
-	char *file = build_pathname("%s", get_amxstring(amx, params[1], 1, len));
-	char *flags = get_amxstring(amx, params[2], 0, len);
+	int length;
+	const char* file  = get_amxstring(amx, params[1], 0, length);
+	const char* flags = get_amxstring(amx, params[2], 1, length);
+
+	FileObject* fp = nullptr;
+
+	if (*params / sizeof(cell) >= 3 && params[3] > 0)
+	{
+		const char* pathID = get_amxstring(amx, params[4], 2, length);
+
+		if (!strcmp(pathID, "ALL"))
+		{
+			pathID = nullptr;
+		}
+
+		fp = ValveFile::Open(file, flags, pathID);
+	}
+	else
+	{
+		fp = SystemFile::Open(build_pathname("%s", file), flags);
+	}
 
-	FILE *fp = fopen(file, flags);
+	if (!fp)
+	{
+		return 0;
+	}
 
-	return (cell)fp;
+	return reinterpret_cast<cell>(fp);
 }
 
+#define BLOCK_INT	4
+#define BLOCK_SHORT	2
+#define BLOCK_CHAR	1
+
+// native fwrite_blocks(file, const data[], blocks, mode);
 static cell AMX_NATIVE_CALL amx_fwrite_blocks(AMX *amx, cell *params)
 {
-	FILE *fp = (FILE *)params[1];
+	FileObject* fp = reinterpret_cast<FileObject*>(params[1]);
 
 	if (!fp)
+	{
 		return 0;
+	}
+
+	cell* data  = get_amxaddr(amx, params[2]);
+	cell blocks = params[3];
+	cell size   = params[4];
 
-	cell *addr = get_amxaddr(amx, params[2]);
-	size_t blocks = params[3];
-	size_t btmp = blocks;
-	cell mode = params[4];
-	switch (mode)
+	size_t read = 0;
+
+	switch (size)
 	{
-	case 1:
+		case BLOCK_CHAR:
 		{
-			char *a = new char[blocks];
-			char *ptr = a;
-			while (btmp--)
-				*ptr++ = static_cast<char>(*addr++);
-			size_t res = fwrite(a, sizeof(char), blocks, fp);
-			delete [] a;
-			return res;
+			for (cell i = 0; i < blocks; ++i)
+			{
+				char value = data[i];
+
+				if (fp->Write(&value, sizeof(value)) != sizeof(value))
+				{
+					break;
+				}
+
+				read += sizeof(value);
+			}
+			break;
 		}
-	case 2:
+		case BLOCK_SHORT:
 		{
-			short *a = new short[blocks];
-			short *ptr = a;
-			while (btmp--)
-				*ptr++ = static_cast<short>(*addr++);
-			size_t res = fwrite(a, sizeof(short), blocks, fp);
-			delete [] a;
-			return res;
+			for (cell i = 0; i < blocks; ++i)
+			{
+				short value = data[i];
+
+				if (fp->Write(&value, sizeof(value)) != sizeof(value))
+				{
+					break;
+				}
+
+				read += sizeof(value);
+			}
+			break;
 		}
-	case 4:
+		default:
```

Comment:
Same here. Invalid parameters were provided, this shouldn't work. If we ever need to add an extra mode here this will get hairy.